### PR TITLE
[HttpClient] Honor "max_duration" when replacing requests with async decorators

### LINKED ISSUE #46316: Max_duration is not respected when retry_count > 1 (http-client)
### Symfony version(s) affected

5.4.8

### Description

Let's say you have an API endpoint that always fails on the first request, and succeed on the second attempt.
Every request takes 30 seconds to be returned by the API.

If you make a Retryable HttpRequest, you can define `timeout` and `max_duration` to 40 seconds.
But it will not be respected, because when doing the second attempt, symfony still waits up to 40 seconds to get a response.

So you can expect this part of code to take 60 seconds to respond with no issues, instead of having a timeout error.

### How to reproduce

1. Create the following `sleep.php` file:
```php
<?php
sleep(30);
$count = (int)@file_get_contents('.count') == 0 ? 1 : (int)@file_get_contents('.count');
if ($count >= 2) {
    unlink('.count');
    echo '2nd attempt. SUCCESS!';
    exit;
}
$count++;
http_response_code(502);
echo 'Failed. ' . $count-1 . ' attempt;';
file_put_contents('.count', (string)$count);
```
2. Run `php -S localhost:8888` in the folder of the file above
3. Create a Command to test it:
```php
<?php
#src/TestCommand.php
namespace App;

use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface;
use Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface;
use Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface;
use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class TestCommand extends \Symfony\Component\Console\Command\Command
{
    protected static $defaultName = 'app:test';
    
    public function __construct(protected HttpClientInterface $httpClient, string $name = null)
    {
        parent::__construct($name);
    }
    
    protected function configure()
    {
        parent::configure(); // TODO: Change the autogenerated stub
    }

    public function execute(InputInterface $input, OutputInterface $output): int
    {
        $output->writeln('Start: ' . date('H:i:s'));
        $start = time();
        $url = 'http://localhost:8888/sleep.php';
        $client = $this->httpClient;
        $response = $client->request('GET', $url, [
            'timeout' => 40,
            'max_duration' => 40
        ]);


        try {
            $output->writeln($response->getStatusCode().': '.$response->getContent(false));
        } catch (ClientExceptionInterface $e) {
            $output->writeln(get_class($e) . ': ' . $e->getMessage());
        } catch (RedirectionExceptionInterface $e) {
            $output->writeln(get_class($e) . ': ' . $e->getMessage());
        } catch (ServerExceptionInterface $e) {
            $output->writeln(get_class($e) . ': ' . $e->getMessage());
        } catch (TransportExceptionInterface $e) {
            $output->writeln(get_class($e) . ': ' . $e->getMessage());
        }
        $output->writeln('Retry coint: ' . (int)$response->getInfo('retry_count'));
        $output->writeln('End: ' . date('H:i:s'));
        $output->writeln('Total time: ' . time()-$start . 's');
        return 1;
    }
}
```

4. Run the test comand with `bin/console app:test` and note that it will succeed on the second attempt after 60 seconds.

### Possible Solution

We may need to work on AmpHttpClient or RetryableHttpClient to use the time left on the other requests.

I posted a half solution [here](https://stackoverflow.com/questions/72194696/symfony-retryable-httpclient-and-timeout-option-how-it-works-when-used-together/72197256#72197256) that may be useful to get the total time of the request, including the waiting time between the attempts.

But it's far from the ideal solution.

### Additional Context

In my case, I have a controller that makes http requests to another endpoint.
The users that call this controller, expect it to return a response in 45 seconds. However, after implementing the RetryableHttpClient, I'm having cases where the request is processed by the other endpoint in the second or third attempt, but my users get a timeout exception.

It happens because the first request to the other endpoint takes less than 45 seconds, but both calls take more than that.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #46316
| License       | MIT
| Doc PR        | -

Instead of #46330

### CODE REVIEW

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/HttpClient/Response/AsyncContext.php [hunk 1]
```diff
@@ -152,13 +153,18 @@ public function getResponse(): ResponseInterface
...
     {
-        $this->info['previous_info'][] = $this->response->getInfo();
+        $this->info['previous_info'][] = $info = $this->response->getInfo();
         if (null !== $onProgress = $options['on_progress'] ?? null) {
             $thisInfo = &$this->info;
             $options['on_progress'] = static function (int $dlNow, int $dlSize, array $info) use (&$thisInfo, $onProgress) {
                 $onProgress($dlNow, $dlSize, $thisInfo + $info);
             };
         }
+        if (0 < ($info['max_duration'] ?? 0) && 0 < ($info['total_time'] ?? 0) && !isset($options['max_duration'])) {
```
Comment (r-martins) [inline src/Symfony/Component/HttpClient/Response/AsyncContext.php]: I'm not sure if I got what you did here, but unless I remove `!` from `!isset($options['max_duration'])` I couldn't make `max_duration` option work as expected when testing it. I will leave more details at the PR's comment area.

**Comment by r-martins** [comment]
Thanks for that, @nicolas-grekas.
I tested your PR with the following test:
```php
$url = 'http://localhost:8888/sleep3.php';
$response = $client->request('POST', $url, [
    'body' => ['foo'=>'bar'],
    'max_duration' => 11
]);
```

Then, the first and the second request fails after 6 seconds and the third would succeed after 5 seconds (total 6+6+5=**17**).

But it turns out that the [part of code where you would change the max_duration](https://github.com/symfony/symfony/pull/46382/files#r875452842) for the retries is never reached, unless the $options['max_duration'] is **not** set. The above request succeeded after around 20 seconds, not triggering the expected timeout error.


If I remove `!`, then yes, I get a timeout error on the further attempts.


If you want to use it, my stupid test script was:

```php
<?php
$count = (int)@file_get_contents('.count') == 0 ? 1 : (int)@file_get_contents('.count');
if ($count <=2) {
    file_put_contents('.count', ++$count);
    sleep(6);
    echo '500';
    http_response_code(500);
    exit;
}

if ($count > 2) {
    sleep(5);
    unlink('.count');
    echo $count . ' attempt. SUCCESS!';
    exit;
}
```
Serve it with php -S and do the test above.

**Comment by r-martins** [comment]
_To think..._
Let's say your `max_duration` is 30 seconds, and the first attempt failed after 29.8 seconds.
Is it worth it to retry with 0.2 second timeout?

We can write another Strategy and deal with it in the `shouldRetry` method. But for most cases, I think that having a `min_duration` or something like that with a reasonable default value would make things easier.

With the current `max_duration` (when working), if we are doing a payment capture request and the second attempt times out after 0.2 second, the payment may have been processed and debited from the customer (after 4 seconds for example), but we would have had a timeout error and asked the customer to try again.

What do you think?

**Review by r-martins — CHANGES_REQUESTED** [review]
Please refer to [this comment](https://github.com/symfony/symfony/pull/46382#issuecomment-1129555088). I didn't know I could review.

File: src/Symfony/Component/HttpClient/Response/AsyncContext.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/AsyncContext.php]: I actually removed the check for `$options['max_duration']`

**Comment by nicolas-grekas** [comment]
Thanks for checking @r-martins! This is now fixed.

> Let's say your max_duration is 30 seconds, and the first attempt failed after 29.8 seconds.
> Is it worth it to retry with 0.2 second timeout?

This reasoning applies to any duration, 0.2 or longer. I don't like the idea of adding a min-duration option because hey, what would be a sensible value for that?!

> With the current max_duration (when working), if we are doing a payment capture request and the second attempt times out after 0.2 second, the payment may have been processed and debited from the customer (after 4 seconds for example), but we would have had a timeout error and asked the customer to try again.

Retries do not happen for non idempotent requests by default. If your strategy retries eg POST requests, then this issue is unrelated to the max_duration as it can happen with any retries. To guard against these, you need to use something like [Stripe's `Idempotency-Key`](https://stripe.com/docs/api/idempotent_requests).

Status: needs review

File: src/Symfony/Component/HttpClient/Response/AsyncContext.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/HttpClient/Response/AsyncContext.php]: is the added test realistic/good enough then? I stared a while at this check but was (wrongly?) reassured by the test reaching the expected exception.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/AsyncContext.php]: testing times is super hard :(

**Review by r-martins — APPROVED** [review]

**Comment by r-martins** [comment]
Thanks for the update @nicolas-grekas. I could test and it's working as expected now.
And yes, my strategy currently retries post requests as the API I'm consuming does not support indempotency-* headers. ;(

Lastly, could you suggest a way to merge your PR in my project? I'm afraid waiting for a new release may take too long. I saw you have it separately, and also http-client is a separated project... Can you advise?

**Pull request merged**